vac/raw/logos-capability-discovery.md

```diff
@@ -0,0 +1,1038 @@
+        else:
+            v ← v.right
+        end if
+    end for
+end procedure
+```
+
+Implementations can extend the IP tree algorithms to IPv6 by using a 128-level binary tree,
+corresponding to the 128-bit length of IPv6 addresses.
+
+### Safety Parameter
+
+The safety parameter `G` ensures waiting times never reach zero even when:
+
+- Service similarity is zero (new service).
+- IP similarity is zero (completely distinct IP)
+
+It prevents ad_cache overflow in cases when attackers try to send ads for random services or from diverse IPs.
+
+### Lower Bound Enforcement
+
+To prevent "ticket grinding" attacks where advertisers repeatedly request new tickets hoping for better waiting times,
+registrars enforce lower bounds:
+
+Invariant: A new waiting time `w_2` at time `t_2`
+cannot be smaller than a previous waiting time `w_1` at time `t_1`
+(where `t_1 < t_2`) by more than the elapsed time:
+
+```text
+w_2 ≥ w_1 - (t_2 - t_1)
+```
+
+Thus registrars maintain lower bound state for:
+
+- Each service in the cache: `bound(s)` and `timestamp(s)`
+- Each IP prefix in the IP tree: `bound(IP)` and `timestamp(IP)`
+
+The total waiting time will respect the lower bound if lower bound is enforced on these.
+These two sets have a bounded size as number of ads present in the ad_cache at a time is bounded by the cache capacity C.
+
+**How lower bound is calculated for service IDs:**
+
+When new service ID `s` enters the cache, `bound(s)` is set to `0`,
+and a `timestamp(s)` is set to the current time.
+When a new ticket request arrives for the same service ID `s`,
+the registrar calculates the service waiting time `w_s` and then applies the lower-bound rule:
+
+`w_s = max(w_s, bound(s) - timestamp(s))`
+
+The values `bound(s)` and `timestamp(s)` are updated whenever a new ticket is issued
+and the condition `w_s > (bound(s) - timestamp(s))`is satisfied.
+
+**How lower bound is calculated for IPs:**
+Registrars enforce lower-bound state for the advertiser’s IP address using IP tree
+(refer to the [IP Similarity Score section](#ip-similarity-score)).
+
+## RPC Messages
+
+All RPC messages are sent using the libp2p Kad-dht message format
+with new message types added for Logos discovery operations.
+
+### Message Types
+
+The following message types are added to the Kad-dht `Message.MessageType` enum:
+
+```protobuf
+enum MessageType {
+    // ... existing Kad-dht message types ...
+    REGISTER = 6;
+    GET_ADS = 7;
+}
+```
+
+### Advertisement Structure
+
+```protobuf
+message Advertisement {
+    // Service identifier (32-byte SHA-256 hash)
+    bytes serviceId = 1;
+
+    // Peer ID of advertiser (32-byte hash of public key)
+    bytes peerId = 2;
+
+    // Multiaddrs of advertiser
+    repeated bytes addrs = 3;
+
+    // Ed25519 signature over (serviceId || peerId || addrs)
+    bytes signature = 4;
+
+    // Optional: Service-specific metadata
+    optional bytes metadata = 5;
+
+    // Unix timestamp in seconds
+    uint64 timestamp = 6;
+}
+```
```

Comment:
I feel like we could probably use signed peer records instead.
They are already implemented and speced.

https://github.com/libp2p/specs/blob/master/RFC/0002-signed-envelopes.md
https://github.com/libp2p/specs/blob/master/RFC/0003-routing-records.md

We already have to modify routing records for random discovery purposes anyway...